enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,782 @@
+If the field is not present, it will be added by the sync controller and defaulted to `machine.openshift.io`.
+
+When the Machine API controller attempts to reconcile an object, it must first check for the presence of the field.
+If it is not present, then the Machine API controller should reconcile the object.
+It is expected that the sync controller will default it to `machine.openshift.io` in this case.
+
+We will implement predicates/event filters for both Machine API and Cluster API controllers.
+
+Machine API controllers will be allowed to reconcile any object that is either:
+
+* Missing the authoritative field in the status (as today)
+* Has the authoritative field set with the `machine.openshift.io` value
+
+Cluster API controllers will be allowed to reconcile any object that either:
+
+* Has a Machine API equivalent AND has the authoritative field set with the `cluster.x-k8s.io` value
+* Does not have a Machine API equivalent
+
+By following the above rules, we should be able to ensure that the authoritative annotation is present in only one location and that the reconciliation of the objects is well defined based on the presence of this annotation.
+
+The APIs will be extended as follows:
+
+```go
+type xxxSpec struct {
+  // authoritativeAPI is the API that is authoritative for this resource.
+  // Valid values are machine.openshift.io and cluster.x-k8s.io.
+  // When set to machine.openshift.io, writes to the spec of the machine.openshift.io copy of this resource will be reflected into the cluster.x-k8s.io copy.
+  // When set to cluster.x-k8s.io, writes to the spec of the cluster.x-k8s.io copy of this resource will be reflected into the machine.openshift.io copy.
+  // Updates to the status will be reflected in both copies of the resource, based on the controller implementing the functionality of the API.
+  // Currently the authoritative API determines which controller will manage the resource, this will change in a future release.
+  // To ensure the change has been accepted, please verify that the `status.authoritativeAPI` field has been updated to the desired value and that the `Synchronized` condition is present and set to `True`.
+  // +kubebuilder:validation:Enum=machine.openshift.io;cluster.x-k8s.io
+  // +kubebuilder:validation:Default:=machine.openshift.io
+  // +default:=machine.openshift.io
+  // +optional
+  AuthoritativeAPI string `json:"authoritativeAPI,omitempty"`
+}
+
+type xxxStatus struct {
+  // authoritativeAPI is the API that is authoritative for this resource.
+  // Valid values are machine.openshift.io and cluster.x-k8s.io.
+  // This value is updated by the sync controller to reflect the authoritative API.
+  // Machine API and Cluster API controllers use this value to determine whether or not to reconcile the resource.
+  // +kubebuilder:validation:Enum=machine.openshift.io;cluster.x-k8s.io
+  // +optional
+  AuthoritativeAPI string `json:"authoritativeAPI,omitempty"`
```

Comment:
The ordering of setting this and whatever label (or other mechanism) in the cluster API variant is very important.  You need to go through a state where neither controller can update to avoid a flow like

1. machineset.machine/A .spec.authoritative=machine
2. machineset.machine/A .spec.authoritative=cluster
3. machineset.cluster/A  label says to manage
4. machine controller creates new machine (a status field check says this is ok)
5. cluster api controller creates new machine (a label check says this ok)
6. sync controller sets machineset.machine/A .status.authoritative=cluster

Not sure if exact ordering is nailed down somewhere, but I don't recall seeing it in excruciating detail

Reply:
We haven't fleshed out the stopping controller in CAPI from reconciling part yet, as we still have a little exploratory work todo. However, the original idea was to have code injected into CAPI that would determine which resource is authoritative by looking up the MAPI version and checking the status. I think now that this is likely to become very complex and the carries a nightmare.

One avenue we are exploring would be to introduce the concept of pausing into MAPI, which already exists in CAPI.

The spec for `authoritativeAPI` still maintains an enum of `machine.openshift.io` and `cluster.x-k8s.io`, where the status version has two additional values `transitioning.to.machine.openshift.io` and `transitioning.to.cluster.x-k8s.io`.

The flow would then look like:
1. machine.spec.authoritativeAPI updated from machine.openshift.io to cluster.x-k8s.io
2. sync controller sets machine.status.authoritativeAPI to transitioning.to.cluster.x-k8s.io
3. Machine API controller sets machine.status.conditions[type=Paused, status=True] (acknowledgement of hand-off)
4. sync controller sets machine.status.authoritativeAPI to cluster.x-k8s.io
5. sync controller removes "paused annotation" from Cluster API resource
6. Cluster API controllers take over reconciliation

The opposite direction should work in the same way, except the paused annotation will be added to the CAPI resource, and that will trigger the paused condition status change as acknowledgement of handover.

The paused annotation will be owned by the sync controller and at admission time updates to it will be denied unless the CAPI resource is already the authoritative API.

We believe we can replicate that in all of the relevant controllers that we need to, without invasive carries in Cluster API.

Cluster API currently is lacking that condition of acknowledgement of pause, but I've started a thread upstream to try and build a contract around that so that it is consistent across providers.

Reply:
How fast moving is this API and how catastrophic is a miss?  Could introduce a 10s delay, resolve 98% of the problem, leaving the list 2% manually resolvable after the fact?

Reply:
The API is generally pretty slow moving, except when scaling up the cluster or scaling down the cluster. In general I expect that most of these transitions will happen in the slow periods when scaling isn't happening.

The worst possible I can think of, would be if someone tried to migrate a machine during a scale up, where perhaps one side could create an instance, but not have time to persist the providerID, the second side could then, on some platforms, create another instance in theory and we leak one. This is mostly a problem for AWS as most other platforms have deterministic naming for instances.

Is there a concern with the above approach or do you think a delay would be in addition to this?